### PR TITLE
Tk rebase aespielberg msspoly dynamics cleaned

### PR DESCRIPTION
**Pull request by aespielberg**
Cleaned version via Twan of attempt at symbolic dynamics.

Still undefined behavior for at least dynamics and get_params

### CODE REVIEW

**Comment by tkoolen** [comment]
Alright, so does this include the changes to Manipulator.dynamics you mention were missing in https://github.com/RobotLocomotion/drake/pull/1202?

File: cmake [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit 890a290b3c9c15385b098d9e9b02dfe4ce308bf3
```
Comment (tkoolen) [inline cmake]: We'll need to undo this change. In your drake directory, type
```
git submodule update --init --recursive
```
then
```
git add cmake
git commit -m "fixed cmake submodule reference"
```

File: systems/plants/@Manipulator/Manipulator.m [hunk 2]
```diff
@@ -36,6 +36,73 @@
       C_times_v = C-G;
     end
     
+    function [xdot,dxdot] = Hdynamics(obj,t,x,u)
```
Comment (tkoolen) [inline systems/plants/@Manipulator/Manipulator.m]: Why was it necessary to add this?

**Comment by aespielberg** [comment]
I didn't think I committed it last night, but it's highlighted as being
added in the change log, so I'm going to go with yes.

-Andy S.

File: systems/plants/@Manipulator/Manipulator.m [hunk 3]
```diff
@@ -36,6 +36,73 @@
...
+        % want to raise for this method, stating that not all outputs were
+        % assigned.  (since I can't write dxdot anymore)
+        [H,C,B,dH,dC,dB] = obj.manipulatorDynamics(q,v);
+        
+        %ANDY CHANGE
+        %Hinv = inv(H);
+        %Hinv = invmsspoly(H);
+
+        if (obj.num_u>0)
+          vdot = (B*u-C);
```
Comment (tkoolen) [inline systems/plants/@Manipulator/Manipulator.m]: The lhs would not be vdot.

File: systems/plants/@Manipulator/Manipulator.m (hunk 2 above)
Comment (aespielberg) [inline systems/plants/@Manipulator/Manipulator.m]: Why was it necessary to add that function?

File: systems/plants/@Manipulator/Manipulator.m [hunk 4]
```diff
@@ -36,6 +36,73 @@
...
+        xdot = [VqInv*v;vdot];
+        dxdot = [...
+          zeros(obj.num_positions,1), matGradMult(dVqInv, v), VqInv, zeros(obj.num_positions,obj.num_u);
+          dvdot];
+      else
+        [H,C,B] = manipulatorDynamics(obj,q,v);
+        if (obj.num_u>0) tau=B*u - C; else tau=-C; end
+        tau = tau + computeConstraintForce(obj,q,v,H,tau,eye(size(H, 1)));
+
+        vdot = tau;
```
Comment (tkoolen) [inline systems/plants/@Manipulator/Manipulator.m]: This makes no sense.

File: systems/plants/@Manipulator/Manipulator.m (hunk 3 above)
Comment (aespielberg) [inline systems/plants/@Manipulator/Manipulator.m]: Hi Twan,

I should have documented more what that is.  Hdynamics is supposed to
return the same thing as dynamics, but with an H^-1 in front.  This was, I
thought, a useful utility function if you wanted to work in msspoly's to
get symbolic dynamics.  It is impossible to get the full dynamics, because
H cannot be inverted and kept as a polynomial expression.  The Hdynamics,
however, as I'm calling it, can still be useful for generating constraints
on dynamics.

-Andy S.

File: systems/plants/@Manipulator/Manipulator.m (hunk 4 above)
Comment (aespielberg) [inline systems/plants/@Manipulator/Manipulator.m]: It is possible that there is an error in it.  I wound up changing courses
and going the dynamics, symbolicized route.  If you would prefer, I can
remove the Hdynamics function.

File: systems/plants/@Manipulator/Manipulator.m [hunk 5]
```diff
@@ -54,7 +121,65 @@
...
         % assigned.  (since I can't write dxdot anymore)
         [H,C,B,dH,dC,dB] = obj.manipulatorDynamics(q,v);
+        
+        params = obj.getParamFrame().getPoly;
+        symp = sym('p',[length(params), 1]); %TODO: p is probably a bad idea here.
+        
+        
+        
+        try
+            H = double(H);
```
Comment (tkoolen) [inline systems/plants/@Manipulator/Manipulator.m]: What happens if q and v are of type TaylorVar or TrigPoly for example? I think this try/catch will break that.

File: systems/plants/@Manipulator/Manipulator.m [hunk 6]
```diff
@@ -54,7 +121,65 @@
...
+        catch
+            if isa(H, 'msspoly')
+                H = msspoly2sym(params, symp, H);
+            end
+            
+        end
+        
+        try
+            C = double(C);
```
Comment (tkoolen) [inline systems/plants/@Manipulator/Manipulator.m]: See above.

File: systems/plants/@Manipulator/Manipulator.m [hunk 7]
```diff
@@ -54,7 +121,65 @@
...
+        catch
+            if isa(C, 'msspoly')
+                C = msspoly2sym(params, symp, C);
+            end
+        end
+        %TODO: B should always be double, right?
+        
+        try
+            B = double(B);
```
Comment (tkoolen) [inline systems/plants/@Manipulator/Manipulator.m]: See above.

File: systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m [hunk 8]
```diff
@@ -804,7 +804,7 @@
 
       model.dirty = false;
 
-      model = createMexPointer(model);
+     % model = createMexPointer(model);
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m]: bad change

File: systems/plants/@RigidBodyManipulator/doKinematics.m [hunk 9]
```diff
@@ -337,10 +337,11 @@
 ret{1} = zeros(16, nq);
 for i = 2 : nb
   body = bodies(i);
-  T_body_to_parent = T{body.parent} \ T{i};
+  %T_body_to_parent = T{body.parent} \ T{i};
+  T_body_to_parent = T{body.parent}^-1 * T{i};
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: Why was this necessary?

File: systems/plants/@RigidBodyManipulator/doKinematics.m [hunk 10]
```diff
@@ -337,10 +337,11 @@
...
   
   dT_body_to_parentdqi = dHomogTrans(T_body_to_parent, S{i}, qdotToV{i});
-  dT_body_to_parentdq = zeros(numel(T{i}), nq) * dT_body_to_parentdqi(1); % to make TaylorVar work better
+  dT_body_to_parentdq = zeros(length(T{i}(:)), nq) * dT_body_to_parentdqi(1); % to make TaylorVar work better
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: let's just do zeros(16, nq) since we know how this in advance.

File: systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m (hunk 8 above)
Comment (aespielberg) [inline systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m]: Hi Twan,

Yes, I agree.   I don't think this should be merged until a better solution
is determined.

I am curious if you have intuition as to the bug, though.  If the
parameters are msspoly, then this gets called and breaks things.  However,
according to Russ, this shouldn't get called when a manipulator has
msspoly's.

I don't know where the logic is for where this if statement is though.  I
tried to find it, but couldn't, so either I missed it, or that line is in
the mex itself (which I didn't want to touch).

My intuition is that the bug happens when a manipulator has some bodies
which are all double, and somebodies which are part msspoly, but I have not
yet been able to confirm it.

-Andy S.

File: systems/plants/@RigidBodyManipulator/doKinematics.m (hunk 9 above)
Comment (aespielberg) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: msspoly does not support \

-Andy S.

File: systems/plants/@RigidBodyManipulator/doKinematics.m [hunk 11]
```diff
@@ -540,19 +541,26 @@
     
     % v gradient: dJdot_times_vi/dv
     [dtwistdv, v_indices] = geometricJacobian(model, kinsol, world, i, world);
-    %     dtwistdv = zeros(6, nv);
-    %     dtwistdv(:, v_indices) = J;
     nv_branch = length(v_indices);
     body_branch_velocity_num = nv_branch - length(body.velocity_num) + 1 : nv_branch;
     djoint_twistdv = zeros(6, nv_branch);
+    if isa(kinsol.J{i}, 'msspoly')
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: can we change these into something like
```
djoint_twistdv = zeros(6, nv_branch) * J{i}(1)
```
It's ugly, but using isa is very slow.

File: systems/plants/@RigidBodyManipulator/geometricJacobian.m [hunk 12]
```diff
@@ -45,8 +45,14 @@
     transformedMotionSubspaces = cellfun(@transformMotionSubspace, ...
       kinsol.T(joint_path), motionSubspaces, num2cell(signs'), 'UniformOutput', ...
       false); % change frame from body to world
-    J = cell2mat(transformedMotionSubspaces);
-    J = transformTwists(inv(kinsol.T{expressed_in}), J); % change frame from world to expressedIn
+  
+
+    for i = 1:1:length(transformedMotionSubspaces)
+       transformedMotionSubspaces{i} = transformedMotionSubspaces{i};
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/geometricJacobian.m]: This doesn't do anything.

File: systems/plants/@RigidBodyManipulator/geometricJacobian.m [hunk 13]
```diff
@@ -45,8 +45,14 @@
...
+    end
+
+    J = cell2matmsspoly(transformedMotionSubspaces);
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/geometricJacobian.m]: does this break the method for non-msspoly inputs?

File: systems/plants/@RigidBodyManipulator/geometricJacobian.m [hunk 14]
```diff
@@ -45,8 +45,14 @@
...
+    J = transformTwists(invmsspoly(kinsol.T{expressed_in}), J); % change frame from world to expressedIn
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/geometricJacobian.m]: does this break the method for non-msspoly inputs?

File: systems/plants/@RigidBodyManipulator/doKinematics.m (hunk 11 above)
Comment (aespielberg) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: If this will work, sure.  I am not 100% sure I understand it but I will
test it when I get a chance.  I was not aware that isa has huge performance
issues - I thought that was the recommended way to do it, but if that's
slow we should definitely rid of it.

-Andy S.

File: systems/plants/@RigidBodyManipulator/inertiasInWorldFrame.m [hunk 15]
```diff
@@ -30,7 +30,9 @@
   if compute_gradients
     dbody_to_world = kinsol.dTdq{i};
     dbody_to_world = dHomogTransInv(body_to_world, dbody_to_world);
-    dAd_world_to_body = dTransformSpatialMotion(world_to_body, eye(6), dbody_to_world, zeros(numel(Ad_world_to_body), nq));
+    %ANDY CHANGE
+    %dAd_world_to_body = dTransformSpatialMotion(world_to_body, eye(6), dbody_to_world, zeros(numel(Ad_world_to_body), nq));
+    dAd_world_to_body = dTransformSpatialMotion(world_to_body, eye(6), dbody_to_world, zeros(length((Ad_world_to_body(:))), nq));
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/inertiasInWorldFrame.m]: zeros(length((Ad_world_to_body(:))), nq) --> zeros(36, nq)

File: systems/plants/@RigidBodyManipulator/inertiasInWorldFrame.m [hunk 16]
```diff
@@ -30,7 +30,9 @@
   if compute_gradients
     dbody_to_world = kinsol.dTdq{i};
     dbody_to_world = dHomogTransInv(body_to_world, dbody_to_world);
-    dAd_world_to_body = dTransformSpatialMotion(world_to_body, eye(6), dbody_to_world, zeros(numel(Ad_world_to_body), nq));
+    %ANDY CHANGE
+    %dAd_world_to_body = dTransformSpatialMotion(world_to_body, eye(6), dbody_to_world, zeros(numel(Ad_world_to_body), nq));
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/inertiasInWorldFrame.m]: get rid of commented out line and ANDY CHANGE

File: systems/plants/@RigidBodyManipulator/geometricJacobian.m (hunk 13 above)
Comment (aespielberg) [inline systems/plants/@RigidBodyManipulator/geometricJacobian.m]: Not in the tests I've made.  It is possible I've missed a corner case, but
I haven't seen it break yet for any double entires.

-Andy S.

File: systems/plants/@RigidBodyManipulator/manipulatorDynamics.m [hunk 17]
```diff
@@ -122,12 +122,16 @@
...
 
+
 for i = 2 : NB
   Ic = crbs{i};
   Si = kinsol.J{i};
   i_indices = manipulator.body(i).velocity_num;
   F = Ic * Si;
   Hii = Si' * F;
+  if isa(Hii, 'msspoly')
+     H = msspoly(H);
```
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/manipulatorDynamics.m]: can we figure out whether H needs to be an msspoly before this loop?

File: systems/plants/RigidBody.m [hunk 18]
```diff
@@ -316,7 +316,9 @@ function dofnum(obj)
       % class. (maybe I should move it up to there?)
       fn = fieldnames(body.param_bindings);
       for i=1:length(fn)
-        body.(fn{i}) = double(subs(body.param_bindings.(fn{i}),poly,pval));
+%         body.(fn{i}) = double(subs(body.param_bindings.(fn{i}),poly,pval));
+        body.(fn{i}) = subs(body.param_bindings.(fn{i}),poly,pval);
+        if (deg(body.(fn{i}))==0) body.(fn{i}) = double(body.(fn{i}));end
```
Comment (tkoolen) [inline systems/plants/RigidBody.m]: does this break non-msspoly inputs?

Comment (aespielberg) [inline systems/plants/RigidBody.m]: Not in my experience.  This line was a recommendation from Russ, fwiw.

-Andy S.

File: systems/plants/RigidBodyElement.m [hunk 19]
```diff
@@ -1,57 +1,59 @@
...
-        body.(fn{i}) = double(subs(body.param_bindings.(fn{i}),poly,pval));
-      end
-    end
-  end
-  
-  methods    
-    function obj = updateBodyIndices(obj,map_from_old_to_new)
-      % intentionally do nothing. overload if necessary
-    end
+    % handles lazy binding of urdf parameters to rigid body element values
```
Comment (tkoolen) [inline systems/plants/RigidBodyElement.m]: please fix indentation (two spaces per indentation level). Also, you may want to change your tab size and indent size to 2 in your matlab preferences (under Matlab > Editor/Debugger > Tab in preferences).

File: systems/plants/RigidBodyFrame.m [hunk 20]
```diff
@@ -69,7 +69,9 @@
 
       % check to see if we have data for this frame
       if ~isempty(obj.param_binding_T)
-        obj.T = double(subs(obj.param_binding_T, poly, pval));
+        %obj.T = double(subs(obj.param_binding_T, poly, pval));
+        obj.T = subs(obj.param_binding_T, poly, pval);
```
Comment (tkoolen) [inline systems/plants/RigidBodyFrame.m]: is this going to break non-msspoly inputs?

File: util/cell2matmsspoly.m [hunk 21]
```diff
@@ -0,0 +1,7 @@
+function output = cell2matmsspoly(C)
+output = C{1};
+for i = 2:1:length(C)
+   output = [output C{i}];
```
Comment (tkoolen) [inline util/cell2matmsspoly.m]: Can you figure out the size of output beforehand? Dynamically growing the size of output will result in a lot of copies --> slow code.

File: util/dHomogTrans.m [hunk 22]
```diff
@@ -32,5 +32,5 @@
 ret(5:7, :) = Rz * qdot_to_omega_x - Rx * qdot_to_omega_z;
 ret(9:11, :) = -Ry * qdot_to_omega_x + Rx * qdot_to_omega_y;
 ret(13:15, :) = R * qdot_to_twist(4:6, :);
-
+ret(16, :) = 1;
```
Comment (tkoolen) [inline util/dHomogTrans.m]: This is wrong. dHomogTrans returns the gradient of a homogeneous transform. The 16th element in a homogeneous transform is the constant 1. The gradient of a constant w.r.t. anything is zero, not one.

File: util/dHomogTransInv.m [hunk 23]
```diff
@@ -20,4 +20,7 @@
 ret = zeros(numel(T), nq);
 ret = setSubMatrixGradient(ret, dinvT_R, 1:3, 1:3, size(T));
 ret = setSubMatrixGradient(ret, dinvT_p, 1:3, 4, size(T));
-end
\ No newline at end of file
+
+ret(16, :) = 1;
```
Comment (tkoolen) [inline util/dHomogTransInv.m]: wrong.

File: util/getSubMatrixGradient.m [hunk 24]
```diff
@@ -25,7 +25,11 @@
 if sorted
   mask = false(M_size);
   mask(rows, cols) = true;
+  try
   dM_submatrix = dM(mask(:), q_indices);
+  catch
+     disp('debug') 
+  end
```
Comment (tkoolen) [inline util/getSubMatrixGradient.m]: please remove.

File: util/dHomogTransInv.m (hunk 23 above)
Comment (aespielberg) [inline util/dHomogTransInv.m]: Oh :-( sorry I thought this was just the Homogeneous Transform.  I will fix
this.

-Andy S.

File: util/invmsspoly.m [hunk 25]
```diff
@@ -0,0 +1,2 @@
+function output = invmsspoly(M)
+    output = M^-1;
```
Comment (tkoolen) [inline util/invmsspoly.m]: how does this work? From the method name I'm assuming that M is an msspoly matrix. But the inverse of a matrix of polynomials is in general not a matrix of polynomials...

File: util/matGradMult.m [hunk 26]
```diff
@@ -15,7 +15,9 @@
   n = size(b,1);
   m = size(A,2);
   k = size(A,1)/n;
-  y = matGradMultMat(sparse(k,n),b,A,sparse(numel(b),m)); % with a minor abuse of notation
+  %ANDY CHANGE
```
Comment (tkoolen) [inline util/matGradMult.m]: please remove

File: util/matGradMultMat.m [hunk 27]
```diff
@@ -19,7 +19,7 @@
 if isnumeric(B)
   B_diag = kron(speye(q),B);
 else
-  B_diag = kron(eye(q),B);
+  B_diag = kronmsspoly(eye(q),B);
```
Comment (tkoolen) [inline util/matGradMultMat.m]: are numeric and msspoly the only two options?

File: util/quadFormGrad.m [hunk 28]
```diff
@@ -4,8 +4,11 @@
 
 nq = size(dX, 2);
 F = P * X;
-dF = matGradMultMat(P, X, sparse(numel(P), nq), dX);
+
+%dF = matGradMultMat(P, X, sparse(numel(P), nq), dX);
```
Comment (tkoolen) [inline util/quadFormGrad.m]: please remove commented out code

File: util/valuecheck.m [hunk 29]
```diff
@@ -19,7 +19,7 @@
     return;
   end
 end
-
+val = double(val);
```
Comment (tkoolen) [inline util/valuecheck.m]: why was this necessary?

File: util/setSubMatrixGradient.m [hunk 30]
```diff
@@ -28,7 +28,9 @@
 else
   M_indices = reshape(1:prod(M_size), M_size);
   M_submatrix_indices = M_indices(rows, cols);
-  dM(M_submatrix_indices(:), q_indices) = dM_submatrix;
+  %dM(M_submatrix_indices(:), q_indices) = dM_submatrix;
+  dM = msspoly(dM); %needed in order to make sure dM is msspoly
+  dM(M_submatrix_indices(:), find(q_indices)) = dM_submatrix;
```
Comment (tkoolen) [inline util/setSubMatrixGradient.m]: are numeric and msspoly the only two options?

**Comment by tkoolen** [comment]
I'm starting to wonder whether it makes sense at all for us to support msspoly dynamics, since you can't get around taking the inverse of the mass matrix, which means you're outside of the class of polynomials. Not sure, but would having parameters by syms make more sense?

**Comment by tkoolen** [comment]
By the way, drake006 is currently the only build server that can be trusted. It reported 68 test failures, so this needs some work.

File: systems/plants/@RigidBodyManipulator/doKinematics.m (hunk 9 above)
Comment (tkoolen) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: but it doesn't support ^-1 either, does it? In any case, the results can't be right in general, as you're outside of the class of polynomials.

**Comment by aespielberg** [comment]
Hi Twan,

I super appreciate this review.

RE: msspoly dynamics, or doing sym - I spoke with Russ and he said that sym
is very slow, which is why msspoly is used.  Since inverses are only
actually required at the end of the calculation, it could make sense to do
as much as possible msspoly, and then convert to sym - which was my current
take on it.  The HDynamics was another take on it - it's not quite the
dynamics, but it's H^-1 * dynamics, which might be useful for many
applications.

If it makes sense to continue down this route, then, on top of the current
problems you've pointed out being fixed, a few things need to happen:

a) I noticed dynamics can't be taken for the case where, x, u, t, etc. are
mspoly, at least not after my changes (was this working prior?).  That is
because of issues with TrigPoly playing well with the code-base.  I imagine
this is might be something that I introduced here.

b) I need to demonstrate that the Hdynamics system makes sense and is
useful.

I think ther ewere other things I had in mind that need to be addressed
before it's merged outside of these issues, but I've lost my train of
thought, so I'll try to keep a list oging.

-Andy S.

File: systems/plants/@RigidBodyManipulator/doKinematics.m (hunk 9 above)
Comment (aespielberg) [inline systems/plants/@RigidBodyManipulator/doKinematics.m]: True - this is one of the first changes I made, but it also made the test
work.  However, it is possible that it worked because of a special case.  I
am unsure, I would have to investigate more carefully on more plants.

If this isn't general, this is something very core, just the simple
kinematics.  If this can't work, then I agree it would probably be best to
scrap the msspoly and consider something symbolic.

-Andy S.

**Comment by tkoolen** [comment]
OK, maybe this can work if you never need the inverse of H. I believe all of the ^-1's you added had homogeneous transforms as their arguments, and homogeneous transforms are very special in that you can compute their inverse using only transposition, matrix multiplication and subtraction. I think if all of the ^-1's are replaced by homogTransInv calls, that part should be OK.

File: systems/plants/@Manipulator/Manipulator.m (hunk 3 above)
Comment (tkoolen) [inline systems/plants/@Manipulator/Manipulator.m]: OK, I understand now. I'm not sure if it makes sense to have a new method in Manipulator for this, as you can already get what you need in two lines when you need it:
```
[~, C, B] = obj.manipulatorDynamics(q, v);
mass_matrix_times_vdot = B * u - C;
```
Getting rid of this method would avoid having to find a better name for it, and to me it seems the use of the method is limited to this msspoly input case, which you can specialize on all you want in your own code.

**Comment by tkoolen** [comment]
Should we close this?

I'm all for supporting msspoly and I may have been a bit harsh on the code review, but honestly there were a lot of bad changes in this pull request.

Should having polynomial/symbolic parameters be a requirement for the C++ implementation?

**Comment by RussTedrake** [comment]
will close for now.  yes -- I definitely think we need to think about how to allow poly/symbolic parameters in the c++ rbm class.  we'll need it soon enough.

**Comment by aespielberg** [comment]
Yeah, close it.  I am looking at finding a better solution right now.
